### PR TITLE
feat: background sync (with threads)

### PR DESCRIPTION
**Pull request by cristianmtr**
Closes https://github.com/jina-ai/internal-tasks/issues/293

Issues

- [x] timestamp timezone difference
- [x] psql connection pool gets exhausted
- [x] locking resources in threaded access

NOTE: Even if we don't merge this, the refactoring of PSQL Handler still needs to be merged, as the previous usage of Conn Pool had issues.

### CODE REVIEW

File: executor/hnswpsql.py [hunk 1]
```diff
@@ -330,3 +336,23 @@ def cleanup(self, **kwargs):
...
             self.logger.warning(f'PSQL has not been initialized')
+
+    def _start_auto_sync(self):
+        self.sync_thread = Thread(target=self._sync_loop, daemon=True)
+        self.sync_thread.start()
+
+    def close(self) -> None:
+        if hasattr(self, 'sync_thread'):
+            try:
+                self.sync_thread.join(1)
```
Comment (JoanFM) [inline executor/hnswpsql.py]: u should not join, u should use some threading Event to communicate that it needs to terminate

Comment (JoanFM) [inline executor/hnswpsql.py]: u should not join, u should use some threading Event to communicate that it needs to terminate

**Review by JoanFM — COMMENTED** [review]
Remember to protect the shared resources, specially when writing.

File: executor/hnswpsql.py [hunk 2]
```diff
@@ -13,6 +15,8 @@
 from .hnswlib_searcher import HnswlibSearcher, DEFAULT_METRIC
 from .postgres_indexer import PostgreSQLStorage
 
+STOP_SYNC_THREAD = False
```
Comment (JoanFM) [inline executor/hnswpsql.py]: do not use a global, use a `threading.Event` or a `flag` as attribute to communicate between threads

**Review by JoanFM — COMMENTED** [review]
Have u tried to put this at the limit? There is no `Locking` mechanism protecting the `vec_index`

File: executor/hnswpsql.py [hunk 3]
```diff
@@ -352,25 +350,23 @@ def _start_auto_sync(self):
...
     def close(self) -> None:
-        global STOP_SYNC_THREAD
-
         if hasattr(self, 'sync_thread'):
-            STOP_SYNC_THREAD = True
+            # wait for sync thread to finish
+            self.stop_sync_thread = True
             try:
                 while self.sync_thread.is_alive():
                     time.sleep(2)
```
Comment (JoanFM) [inline executor/hnswpsql.py]: simply use join

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: executor/hnswpsql.py (hunk 3 above)
Comment (cristianmtr) [reply executor/hnswpsql.py]: didn't you say not to use join and use a flag?

Comment (JoanFM) [reply executor/hnswpsql.py]: u use join but not with timeout. U use join to wait for the flag to take its effects. The problem was using join with a timeout setting

Comment (cristianmtr) [reply executor/hnswpsql.py]: ah I see. Fixed

**Review by JoanFM — CHANGES_REQUESTED** [review]
Lock should be used as context manager, otherwise it is very dangerous and can lead to deadlocks

**Comment by cristianmtr** [comment]
> Lock should be used as context manager, otherwise it is very dangerous and can lead to deadlocks

Yes, done

**Review by JoanFM — CHANGES_REQUESTED** [review]
It is not clear to me which is the shared resource, is it vec indexer or also kv indexer? and why are them not both protected?

If vec indexer is the shared resource, then u can optimize much better to have the resource lock much less time

File: executor/hnswpsql.py [hunk 4]
```diff
@@ -3,7 +3,11 @@
 
 import copy
 import inspect
-from datetime import datetime
+import threading
+import time
+import traceback
```
Comment (JoanFM) [inline executor/hnswpsql.py:8]: is this import needed?

File: executor/hnswpsql.py [hunk 5]
```diff
@@ -156,34 +171,50 @@ def _sync(
...
+                shard_id=self.runtime_args.pea_id,
+                total_shards=self.total_shards,
+                timestamp=timestamp,
+            )
 
-        if rebuild or self._vec_indexer.size == 0:
-            # call with just indexing
-            self._vec_indexer = HnswlibSearcher(**self._init_kwargs)
-            self._vec_indexer.index_sync(iterator, batch_size)
+            if rebuild or self._vec_indexer.size == 0:
```
Comment (JoanFM) [inline executor/hnswpsql.py:189]: u can move vec index lock aquiring here.

And maybe have another lock for the kv indexer

File: executor/hnswpsql.py [hunk 6]
```diff
@@ -156,34 +171,50 @@ def _sync(
...
             else:
-                self.logger.error(
-                    f'No timestamp provided in parameters: '
-                    f'and vec_indexer.last_timestamp'
-                    f'was None. Cannot do sync'
-                )
-                return
-        else:
-            timestamp = datetime.fromisoformat(timestamp)
+                timestamp = datetime.fromisoformat(timestamp)
```
Comment (maximilianwerk) [inline executor/hnswpsql.py]: Can you put this into a separate function? Helps readability.

File: tests/integration/test_sync.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        while i < nr_runs:
+            docs = get_documents(
+                nr=nr_docs_batch, index_start=i * nr_docs_batch, emb_size=emb_size
+            )
+
+            f.post('/index', docs)
+
+            time.sleep(10)  # wait for syncing
+
+            result = f.post('/search', search_docs, return_results=True)
```
Comment (maximilianwerk) [inline tests/integration/test_sync.py]: Can we add a test that does something along:
```
found_something = False
for i in range(100):
    result = f.post(...)
    if results[...]: # check if results are filled
        found_something = True
    time.sleep(0.1)
assert found_something
```
In order to see, that we actually can send search requests all the time?

File: tests/integration/test_sync.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    search_docs = DocumentArray(
+        get_documents(index_start=nr_docs_batch * (nr_runs + 1), emb_size=emb_size)
+    )
+
+    f = Flow().add(
+        uses=HNSWPostgresIndexer,
+        uses_with=uses_with,
+    )
+
+    with f:
```
Comment (maximilianwerk) [inline tests/integration/test_sync.py:42]: Do we need a Flow for this? Or could we just do it on the Executor object?

File: executor/hnswpsql.py (hunk 5 above)
Comment (JoanFM) [reply executor/hnswpsql.py:189]: I think `kv indexer` is safe if `_get_delta` does not change the object

File: executor/hnswpsql.py [hunk 9]
```diff
@@ -156,34 +171,50 @@ def _sync(
         batch_size: int = 100,
         **kwargs,
     ):
-        if timestamp is None:
-            if rebuild:
-                timestamp = datetime.min
-            elif self._vec_indexer.last_timestamp:
-                timestamp = self._vec_indexer.last_timestamp
+        # we prevent race conditions with search
+        with self.lock:
```
Comment (JoanFM) [inline executor/hnswpsql.py]: Definitely, the lock can be moved after `_get_delta` so that while get_delta happens u can still do search

Comment (cristianmtr) [reply executor/hnswpsql.py]: The lock needs to guard the vec_indexer's state. I moved it to better fit that. It guards it at search time and guards at sync time.

File: tests/integration/test_sync.py (hunk 8 above)
Comment (cristianmtr) [reply tests/integration/test_sync.py:42]: We could, but doing it in a Flow is more "realistic"

File: tests/integration/test_sync.py (hunk 7 above)
Comment (cristianmtr) [reply tests/integration/test_sync.py]: Not sure I get how this suggestion fits. You mean keep searching without indexing new docs? Or do you mean waiting less than 10 secs?

Comment (maximilianwerk) [reply tests/integration/test_sync.py]: Exactly, don't wait but search the whole time and at some point the search should start returning better results.

File: executor/hnswpsql.py (hunk 6 above)
Comment (cristianmtr) [reply executor/hnswpsql.py]: Done

File: tests/integration/test_sync.py (hunk 8 above)
Comment (maximilianwerk) [reply tests/integration/test_sync.py:42]: Yes, I am OK with having an integration test, that tests with the Flow. But not each and every test must test the whole environment. Until recently we had the necessity to test on Flow level. But now this is gone and testing on lower level makes the tests less complex. This helps others understanding, what actually happens.

Comment (cristianmtr) [reply tests/integration/test_sync.py:42]: > But now this is gone

Why is the necessity gone now?

File: tests/integration/test_sync.py (hunk 7 above)
Comment (cristianmtr) [reply tests/integration/test_sync.py]: Got it. Check again

File: tests/integration/test_sync.py (hunk 8 above)
Comment (maximilianwerk) [reply tests/integration/test_sync.py:42]: My understanding was, that until recently `MyExecutor()` would not initialize all the implicit arguments that usually would be initialized by the runtime. But now it does and thus `MyExecutor()` for local testing is a valid way to go. Might be, that my understanding here is wrong.

Comment (cristianmtr) [reply tests/integration/test_sync.py:42]: Yes, you can fake them by passing runtime_args and so on as args to the `__init__`. But I'd say testing with the full Flow is the scenario we want to showcase and make sure runs as expected.

File: executor/hnswpsql.py (hunk 4 above)
Comment (cristianmtr) [reply executor/hnswpsql.py:8]: Yes, so we can get a stack trace when the thread fails.

Comment (cristianmtr) [reply executor/hnswpsql.py:8]: Maybe we need to subclass Thread as we did once in the core to better handle exceptions. That should be another PR. This one is big enough as is.

File: executor/hnswpsql.py (hunk 5 above)
Comment (cristianmtr) [reply executor/hnswpsql.py:189]: Yes, that's what I did. Can you check again?

**Review by JoanFM — APPROVED** [review]

**Pull request merged**